cpp/src/utilities/wrapper_types.hpp

```diff
@@ -291,8 +292,78 @@ CUDA_HOST_DEVICE_CALLABLE
 {
   return value;
 }
+
+/**---------------------------------------------------------------------------*
+ * @brief Trait to use to get underlying type of wrapped object
+ * 
+ * This struct can be used with either a fundamental type or a wrapper type and
+ * it uses unwrap to get the underlying type.
+ * 
+ * Example use case: 
+ *  Making a functor to use with a `type_dispatcher` that works on the
+ *  underlying type of all `gdf_dtype`
+ *  
+ * ```c++
+ * struct example_functor{
+ *  template <typename T>
+ *  int operator()(){
+ *    using T1 = cudf::detail::unwrapped_type<T>::type;
+ *    return sizeof(T1);
+ *  }
+ * };
+ * ```
+ * 
+ * @tparam T Either wrapped object type or fundamental type
+ *---------------------------------------------------------------------------**/
+template <typename T>
+struct unwrapped_type {
+  using type = std::decay_t<decltype(unwrap(std::declval<T&>()))>;
+};
+
+/**---------------------------------------------------------------------------*
+ * @brief Helper type for `unwrapped_type`
+ * 
+ * Example:
+ * ```c++
+ * using T1 = cudf::detail::unwrapped_type_t<date32>; // T1 = int 
+ * using T2 = cudf::detail::unwrapped_type_t<float>;  // T2 = float 
+ * ```
+ * 
+ * @tparam T Either wrapped object type or fundamental type
+ *---------------------------------------------------------------------------**/
+template <typename T>
+using unwrapped_type_t = typename unwrapped_type<T>::type;
+
 } // namespace detail
 
+/**---------------------------------------------------------------------------*
+ * @brief Clone of std::numeric_limits that also works on wrapper types
+ * 
+ * @tparam T 
+ *---------------------------------------------------------------------------**/
+template <typename T>
+struct numeric_limits {
```

Comment:
If I understand correctly, an alternative approach would be to provide a specialization of `std::numeric_limits` for our wrapper types rather than providing a new numeric_limits class...  There is debate about whether doing so is legal in C++11, but it seems to me that it is legal to specialize `std::numeric_limits` for non-standard types according to this post: https://stackoverflow.com/questions/16122912/is-it-ok-to-specialize-stdnumeric-limitst-for-user-defined-number-like-class

Did you consider this approach? It would avoid violating the opaqueness of wrapper types that @jrhemstad wants to preserve, instead defining semantics for min() and max() of those wrapper types.

Reply:
I did consider and saw the same link you posted but considering that the other answer advised against this approach, I thought I'm not aware enough about the standards to make a call, so I chose the safer option. Although that option is quite appealing.

Reply:
ok, so I took your advice https://github.com/rapidsai/cudf/pull/1076/commits/d18f8dc9d459b0aaf993e06a69b39f5672d4b4aa and nothing seems to break.